File: rsnmm.c
#include <R.h>
#include <Rmath.h>
#include <R_ext/Lapack.h>

#define expit(a) (exp(a) / (1 + exp(a)))

void
rsnmm(int *size, int *tmax,
      double *ty, double *tmod, double *tavail, double *tstate,
      double *beta, double *eta, double *mu, double *theta,
      double *coefavail,  double *coefstate, double *coeferr,
      int *avail, double *base, int *state, int *a, double *prob,
      double *y, double *err, double *statec, double *ac, double *availc)
{
  int i, j, n = *size, T = *tmax;
  double r, q, ym;
  GetRNGstate();
  for (i = 0; i < n; i++) {
    for (j = 1; j < T; j++) {
      /* probability of availabilty */
      r = expit(coefavail[0]
                + coefavail[1] * tavail[j]
                + coefavail[2] * a[i*T + j - 1]
                + coefavail[3] * y[i*T + j - 1]);
      /* availability - uncentered and centered */
      avail[i*T + j] = (int) rbinom(1, r);
      availc[i*T + j] = avail[i*T + j] - r;
      /* probability that binary state is +1 */
      q = expit(coefstate[0]
                + coefstate[1] * tstate[j]
                + coefstate[2] * base[i*T + j - 1]
                + coefstate[3] * state[i*T + j - 1]
                + coefstate[4] * a[i*T + j - 1]);
      /* binary state on {-1, 1} - uncentered and centered */
      state[i*T + j] = rbinom(1, q) < 1 ? -1 : 1;
      statec[i*T + j] = state[i*T + j] - (q - (1 - q));
      /* treatment probability */
      prob[i*T + j] = avail[i*T + j] * expit(eta[0]
                                          + eta[1] * base[i*T + j]
                                          + eta[2] * state[i*T + j]
                                          + eta[3] * a[i*T + j - 1]
                                          + eta[4] * y[i*T + j - 1]);
      /* treatment indicator - uncentered and centered */
      a[i*T + j] = (int) rbinom(1, prob[i*T + j]);
      ac[i*T + j] = a[i*T + j] - prob[i*T + j];
      /* conditional mean response */
      ym = mu[0]
        + mu[1] * ty[j]  /* pre-evaluated time function */
        + mu[2] * base[i*T + j]
        + ac[i*T + j] * (beta[0]
                         + beta[1] * tmod[j] /* pre-evaluated time function */
                         + beta[2] * base[i*T + j]
                         + beta[3] * state[i*T + j]
                         + beta[4] * a[i*T + j - 1])
        + ac[i*T + j - 1] * (beta[5]
                             + beta[6] * tmod[j - 1]
                             + beta[7] * base[i*T + j - 1]
                             + beta[8] * state[i*T + j - 1])
        + theta[0] * availc[i*T + j]
        + theta[1] * statec[i*T + j]
        + theta[2] * availc[i*T + j - 1]
        + theta[3] * statec[i*T + j - 1];
      /* error */
      err[i*T + j] += *coeferr * err[i*T + j - 1];
      /* response */
      y[i*T + j] = ym + err[i*T + j];
    }
  }
  PutRNGstate();
  return;
}

File: rsnmm.R
rsnmm <- function(n, tmax, control, ...) {
  control <- if (missing(control)) rsnmm.control(...)
             else do.call("rsnmm.control", control)
  tmax <- tmax + (tmax %% 2) + 1
  time <- rep(0:(tmax - 1), n)
  tfun <- do.call("data.frame", lapply(control$tfun, function(f) f(time, tmax)))
  coef.err <- 0
  control$cormatrix <- matrix(control$coralpha, tmax, tmax)
  diag(control$cormatrix) <- 1
  if (control$corstr == "exchangeable") {
    err <- sqrt(control$coralpha) * rep(rnorm(n, sd = control$sd), each = tmax)
    err <- err + rnorm(n * tmax, sd = sqrt(with(control, sd^2 * (1 - coralpha))))
  }
  else {
    ## provisional error
    err <- ifelse(time == 0, rnorm(n, sd = control$sd),
                  rnorm(n * (tmax - 1),
                        sd = sqrt(with(control, sd^2 * (1 - coralpha^2)))))
    err[time == 0] <- rnorm(n, sd = control$sd)
    coef.err <- control$coralpha
    control$cormatrix <- matrix(with(control,
                                     coralpha^(abs(row(cormatrix) -
                                                   col(cormatrix)))), tmax, tmax)
  }
  d <- .C("rsnmm",
          n = as.integer(n),
          tmax = as.integer(tmax),
          ty = as.double(tfun$ty),
          tmod = as.double(tfun$tmod),
          tavail = as.double(tfun$tavail),
          tstate = as.double(tfun$tstate),
          beta = with(control, as.double(c(beta0, beta1))),
          eta = as.double(control$eta),
          mu = as.double(control$mu),
          theta = with(control, as.double(c(theta0, theta1))),
          coef.avail = as.double(control$coef.avail),
          coef.state = as.double(control$coef.state),
          coef.err = as.double(coef.err),
          avail = as.integer(rep(0, n * tmax)),
          base = as.double(rep(rnorm(n), each = tmax)),
          state = as.integer(rep(0, n * tmax)),
          a = as.integer(rep(0, n * tmax)),
          prob = as.double(rep(0, n * tmax)),
          y = as.double(rep(0, n * tmax)),
          err = as.double(err),
          state.center = as.double(rep(0, n*tmax)),
          a.center = as.double(rep(0, n*tmax)),
          avail.center = as.double(rep(0, n*tmax)))
  d <- data.frame(id = rep(1:n, each = tmax), time = time,
                  ty = d$ty, tmod = d$tmod, tavail = d$tavail, tstate = d$tstate,
                  base = d$base, state = d$state, a = d$a, y = d$y, err = d$err,
                  avail = d$avail, prob = d$p, a.center = d$a.center,
                  state.center = d$state.center, avail.center = d$avail.center,
                  one = 1)
  ## nb: for a given row, y is the proximal response
  d$lag1y <- with(d, delay(id, time, y))
  d$lag2y <- with(d, delay(id, time, y, 2))
  d$lag1err <- with(d, delay(id, time, err))
  d$lag1avail <- with(d, delay(id, time, avail))
  d$lag1avail.center <- with(d, delay(id, time, avail.center))
  d$lag2avail <- with(d, delay(id, time, avail, 2))
  d$lag2avail.center <- with(d, delay(id, time, avail.center, 2))
  d$lag1a <- with(d, delay(id, time, a))
  d$lag2a <- with(d, delay(id, time, a, 2))
  d$lag1prob <- with(d, delay(id, time, prob))
  d$lag2prob <- with(d, delay(id, time, prob, 2))
  d$lag1a.center <- with(d, delay(id, time, a.center))
  d$lag2a.center <- with(d, delay(id, time, a.center, 2))
  d$lag1tmod <- with(d, delay(id, time, tmod))
  d$lag2tmod <- with(d, delay(id, time, tmod, 2))
  d$lag1state <- with(d, delay(id, time, state))
  d$lag1state.center <- with(d, delay(id, time, state.center))
  rownames(d) <- NULL
  attributes(d) <- c(attributes(d), control)
  d
}

rsnmm.control <- function(origin = 1, sd = 1,
                          coralpha = sqrt(0.5),
                          corstr = c("ar1", "exchangeable"),
                          beta0 = c(-0.2, 0, 0, 0.2, 0), beta1 = rep(0, 4),
                          eta = c(0, 0, 0.8, -0.8, 0), mu = rep(0, 3),
                          theta0 = c(0, 0.8), theta1 = c(0, 0),
                          coef.avail = c(100, rep(0, 3)), coef.state = rep(0, 5),
                          tfun = NULL, lag = 3 + any(beta1 != 0)) {
  corstr <- match.arg(corstr)
  if (is.null(tfun))
    tfun <- rep(list(function(tcur, tmax) rep(0, length(tcur))), 4)
  list(origin = 1, lag = lag,
       ## error SD, correlation
       sd = sd, coralpha = coralpha, corstr = corstr,
       ## proximal effect coefficients
       beta0 = setNames(beta0, c("one", "tmod", "base", "state", "lag1a")),
       ## delayed effect coefficients
       beta1 = setNames(beta1, c("one", "lag1tmod", "base", "lag1state")),
       ## treatment probability model coefficients
       eta = setNames(eta, c("one", "base", "state", "lag1a", "lag1y")),
       ## exogenous or time-invariant main effects
       mu = setNames(mu, c("one", "ty", "base")),
       ## time-varying main effects, centered and proximal
       theta0 = setNames(theta0, c("avail", "state")),
       ## time-varying main effects, centered and delayed
       theta1 = setNames(theta1, c("lag1avail", "lag1state")),
       ## availability model coefficients
       coef.avail = setNames(coef.avail, c("one", "tavail", "lag1a", "lag1y")),
       ## binary state model coefficients
       coef.state = setNames(coef.state,
                            c("one", "tstate", "base", "lag1state", "lag1a")),
       ## functions of time in the main effect, proximal effect,
       ## availability model, and binary state model
       tfun = setNames(tfun, c("ty", "tmod", "tavail", "tstate")))
}
